subprojects/dependency-management/src/test/groovy/org/gradle/internal/component/model/AttributePrecedenceSchemaAttributeMatcherTest.groovy

```diff
@@ -109,10 +109,25 @@ class AttributePrecedenceSchemaAttributeMatcherTest extends Specification {
         matcher.match(selectionSchema, [candidate6], requested, null, explanationBuilder) == [candidate6]
     }
 
+    def "disambiguates extra attributes in precedence order"() {
+        def candidate1 = candidate("best", null, "a")
+        def candidate2 = candidate("another", "best", "b")
+        def candidate3 = candidate("another", "another", "c")
+
+        expect:
+        matcher.match(selectionSchema, [candidate1, candidate2, candidate3], requested("best", null, null), null, explanationBuilder) == [candidate1]
+        matcher.match(selectionSchema, [candidate1, candidate2, candidate3], requested("another", null, null), null, explanationBuilder) == [candidate2]
+        matcher.match(selectionSchema, [candidate1, candidate2, candidate3], requested("another", "another", null), null, explanationBuilder) == [candidate3]
+    }
```

Comment:
This doesn't look quite right to me. We're changing how the compatibility rule works and the first and third tests don't use the precedence ordering at all.

I think part of the problem is that you're not really running into a lot of extra attributes because the requested query overlaps with the candidates too much.

I would try something like this:
```
    def "disambiguates extra attributes in precedence order"() {
        def candidate1 = candidate("best", "compatible", "compatible")
        def candidate2 = candidate("compatible", "best", "compatible")
        def candidate3 = candidate("compatible", "compatible", "best")
        def candidate4 = candidate("compatible", "compatible", "compatible")
        def requested = AttributeTestUtil.attributes("unknown": "unknown")

        expect:
        matcher.match(selectionSchema, [candidate1, candidate2, candidate3, candidate4], requested, null, explanationBuilder) == [candidate1]
        matcher.match(selectionSchema, [candidate2, candidate3, candidate4], requested, null, explanationBuilder) == [candidate2]
        matcher.match(selectionSchema, [candidate3, candidate4], requested, null, explanationBuilder) == [candidate3]
    }
```

When given different sets of candidates, we can see that we pick the "best" candidate based on precedence order with the same requested attributes each time.

I also removed the change to the AttributeCompatibilityRule above too.